mesh/spacing.hpp

```diff
@@ -0,0 +1,689 @@
+   inline int Size() const { return n; }
+
+   /// Sets the size, or number of intervals (elements).
+   virtual void SetSize(int size) = 0;
+
+   /// Sets the property that determines whether the spacing is reversed.
+   void SetReverse(bool r) { reverse = r; }
+
+   /// Returns the width of interval @a p (between 0 and @a Size() - 1).
+   virtual double Eval(int p) = 0;
+
+   /// Returns the width of all intervals, resizing @a s to @a Size().
+   void EvalAll(Vector & s)
+   {
+      s.SetSize(n);
+      for (int i=0; i<n; ++i)
+      {
+         s[i] = Eval(i);
+      }
+   }
+
+   /** @brief Scales parameters by the factor @a a associated with @a Size().
+
+       Note that parameters may be scaled inversely during coarsening and
+       refining, so the scaling should be linear in the sense that scaling by a
+       number followed by scaling by its inverse has no effect on parameters. */
+   virtual void ScaleParameters(double a) { }
+
+   /// Returns the spacing type, indicating the derived class.
+   virtual SPACING_TYPE SpacingType() const = 0;
+
+   /** @brief Prints all the data necessary to define the spacing function and
+       its current state (size and other parameters).
+
+       The format is generally
+       SPACING_TYPE numIntParam numDoubleParam {int params} {double params} */
+   virtual void Print(std::ostream &os) const = 0;
+
+   /// Returns the number of integer parameters defining the spacing function.
+   virtual int NumIntParameters() const = 0;
+
+   /// Returns the number of double parameters defining the spacing function.
+   virtual int NumDoubleParameters() const = 0;
+
+   /// Returns the array of integer parameters defining the spacing function.
+   /// @param[out] p  Array of integer parameters, resized appropriately.
+   virtual void GetIntParameters(Array<int> & p) const = 0;
+
+   /// Returns the array of double parameters defining the spacing function.
+   /// @param[out] p  Array of double parameters, resized appropriately.
+   virtual void GetDoubleParameters(Vector & p) const = 0;
+
+   /// Returns true if the spacing function is nested during refinement.
+   virtual bool Nested() const = 0;
+
+   /// Returns a clone of this spacing function.
+   virtual SpacingFunction *Clone() const;
+
+   virtual ~SpacingFunction() { }
+
+protected:
+   int n;  // Size, or number of intervals (elements)
+   bool reverse;  // Whether to reverse the spacing
+   bool scale;  // Whether to scale parameters in ScaleParameters.
+};
+
+/** @brief Uniform spacing function, dividing the unit interval into @a Size()
+    equally spaced intervals (elements).
+
+    This function is nested and has no scaled parameters. */
+class UniformSpacingFunction : public SpacingFunction
+{
+public:
+   UniformSpacingFunction(int n_)
+      : SpacingFunction(n_)
+   {
+      CalculateSpacing();
+   }
+
+   virtual void SetSize(int size) override
+   {
+      n = size;
+      CalculateSpacing();
+   }
```

Comment:
`virtual` is superfluous on all these `override` methods.

Reply:
Removed.